rust/core-lib/src/file.rs

```diff
@@ -250,32 +250,62 @@ where P: AsRef<Path>
         .ok()
 }
 
-impl From<io::Error> for FileError {
-    fn from(src: io::Error) -> FileError {
-        FileError::Io(src)
-    }
-}
-
-impl From<str::Utf8Error> for FileError {
-    fn from(_: str::Utf8Error) -> FileError {
-        FileError::UnknownEncoding
-    }
-}
-
 impl From<FileError> for RemoteError {
-    fn from(_src: FileError) -> RemoteError {
+    fn from(src: FileError) -> RemoteError {
         //TODO: when we migrate to using the failure crate for error handling,
         // this should return a better message
-        RemoteError::custom(5, "failed to load file", None)
+		use std::error::Error;
+		let mut code = 0;
+		let mut message = String::new();
+		match src {
+			FileError::Io(e, p) => {
+				code = 5;
+				match p.to_str() {
+					Some(s) => {
+						message.push_str(e.description());
+						message.push_str(" ");
+						message.push_str(s);
+					},
+					None => {
+						message.push_str(e.description());
+					}
+				};
+			},
+			FileError::UnknownEncoding(p) => {
+				code = 6;
+				match p.to_str() {
+					Some(s) => {
+						message.push_str("Error decoding file: ");
+						message.push_str(s);
+					},
+					None => {
+						message.push_str("Error decoding file");
+					}
+				};
+			},
+			FileError::HasChanged(p) => {
+				code = 7;
+				match p.to_str() {
+					Some(s) => {
+						message.push_str("File has changed on disk: ");
+						message.push_str(s);
+					},
+					None => {
+						message.push_str("File has changed on disk");
+					}
+				};
+			}
+		};
+		RemoteError::custom(code, message, None)
     }
 }
 
 impl fmt::Display for FileError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
-            &FileError::Io(ref e) => write!(f, "{}", e),
-            &FileError::UnknownEncoding => write!(f, "Error decoding file"),
-            &FileError::HasChanged => write!(f, "File has changed on disk. \
+            &FileError::Io(ref e, ref _p) => write!(f, "{}", e),
+            &FileError::UnknownEncoding(ref _p) => write!(f, "Error decoding file"),
```

Comment:
So we can just use the path here, in the `write!` macro, and we don't have to worry about string conversion etc because path implements `Debug`, so you can just do `write!(f, "Error decoding file: {:?}", p)` (where `p` is the path) and it will print correctly.